File: 0049-group-anagrams/0049-group-anagrams.py
from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_groups = defaultdict(list)

        for s in strs:
            sorted_s = ''.join(sorted(s))
            anagram_groups[sorted_s].append(s)

        result = list(anagram_groups.values())
        return result

File: 0049-group-anagrams/0049-group-anagrams.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagramGroups = new HashMap<>();
        
        for (String s : strs) {
            char[] charArray = s.toCharArray();
            Arrays.sort(charArray);
            String sortedS = new String(charArray);
            
            if (!anagramGroups.containsKey(sortedS)) {
                anagramGroups.put(sortedS, new ArrayList<>());
            }
            
            anagramGroups.get(sortedS).add(s);
        }
        
        return new ArrayList<>(anagramGroups.values());
    }
}